### PR TITLE
Implement faster JSON decoder compatible with original one

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+            case INTERVAL_YEAR_TO_MONTH:
+            case INTERVAL_DAY_TO_SECOND:
+            case IPADDRESS:
+            case UUID:
+            case DECIMAL:
+            case CHAR:
+            case GEOMETRY:
+            case SPHERICAL_GEOGRAPHY:
+                return new StringDecoder();
+            default:
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:127]: We should encode the physical type in the response metadata so that the client can decode the data even if it doesn't know about the type.

**Review by electrum — COMMENTED** [review]
Can you add a test that has some literal JSON, so that it's easy to see what the JSON looks like?

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java (hunk 1 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:127]: I scheduled a meeting on friday to discuss that

Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:127]: That's what I was going to propose that we should add an SPI that would describe the logical type using physical, on the wire types (like basic primitives, row, array and map).

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+public class JsonQueryDataAccess
+        implements QueryDataAccess
+{
+    private final InputStream stream;
+    private final TypeDecoder[] decoders;
+
+    public JsonQueryDataAccess(List<Column> columns, InputStream stream)
+    {
+        this.decoders = createTypeDecoders(columns);
```
Comment (pettyjamesm) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:45]: Is this creation necessary per instance? Seems like these are fixed for all segments and could be passed directly.

Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:45]: This is WiP still

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java (hunk 1 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:127]: This is for backward compatibility with the existing format. I'll add a new v2 soon that will be dedicated for spooling protocol.

File: client/trino-client/src/main/java/io/trino/client/QueryDataDecoder.java [hunk 3]
```diff
@@ -29,6 +29,16 @@ interface Factory
         String encoding();
     }
 
+    /**
+     * Decodes the input stream into a QueryDataAccess object.
+     * <p></p>
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/QueryDataDecoder.java:34]: Use single `<p>` for Javadoc

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public final class JsonDecodingUtils
+{
+    private JsonDecodingUtils() {}
+
+    static TypeDecoder[] createTypeDecoders(List<Column> columns)
+    {
+        requireNonNull(columns, "columns is null");
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:73]: No need for this check as the next line will throw

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        TypeDecoder[] decoders = new TypeDecoder[columns.size()];
+        int index = 0;
+        for (Column column : columns) {
+            decoders[index++] = createTypeDecoder(column.getTypeSignature());
+        }
+        return decoders;
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:79]: Why not
```java
return columns.stream()
        .map(column -> createTypeDecoder(column.getTypeSignature()))
        .toArray(TypeDecoder[]::new);

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java [hunk 6]
```diff
@@ -13,36 +13,124 @@
...
+                switch (parser.nextToken()) {
+                    case END_ARRAY:
+                        closed = true;
+                        stream.close();
+                        break;
+                    case START_ARRAY:
+                        closed = false;
+                        break;
+                    default:
+                        throw new IllegalStateException("Expected end of array or start of array, but got " + parser.currentToken());
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:99]: Throwing `JsonProcessingException` seems more appropriate

Or throw `VerifyException` as that matches the `verify()` above

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java [hunk 7]
```diff
@@ -13,36 +13,124 @@
...
+    private final TypeDecoder[] decoders;
 
     public JsonQueryDataAccess(List<Column> columns, InputStream stream)
     {
-        this.columns = requireNonNull(columns, "columns is null");
+        this.decoders = createTypeDecoders(columns);
         this.stream = requireNonNull(stream, "stream is null");
     }
 
+    @SuppressModernizer
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:49]: Why is this needed here?

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+        {
+            switch (parser.nextToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_NUMBER_INT:
+                    return parser.getLongValue();
+                case VALUE_STRING:
+                    return Long.parseLong(parser.getText());
+                default:
+                    return illegalToken(parser);
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:147]: The typical way is to do `throw illegalToken(parser)` so that the compiler knows this always throws

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+                    return null;
+                default:
+                    return illegalToken(parser);
+            }
+            List<Object> values = new ArrayList<>(); // nulls allowed
+            while (true) {
+                try {
+                    values.add(typeDecoder.decode(parser));
+                }
+                catch (IllegalTokenException e) {
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:339]: Using exceptions for control flow here seems bad. Also, it's expensive, because we construct the exception with a formatted message and stack trace.

What if we change the flow so that `TypeDecoder` expects a current token in the `JsonParser`? Then `ArrayDecoder` would know to stop when `nextToken() == END_ARRAY` and wouldn't call the child decoder?

**Review by electrum — COMMENTED** [review]
Looks good overall

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java (hunk 7 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:49]: It's not. Removed

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java (hunk 8 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:147]: makes sense

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java (hunk 9 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:339]: Yeah, that makes more sense. I've switched to the assumption that type decoder always moves to the current token to be decoded.

File: client/trino-client/src/main/java/io/trino/client/QueryDataDecoder.java [hunk 10]
```diff
@@ -28,7 +29,17 @@ interface Factory
         String encoding();
     }
 
-    Iterable<List<Object>> decode(InputStream input, DataAttributes segmentAttributes)
+    /**
+     * Decodes the input stream into a QueryDataAccess object.
+     * <p>
+     * Closing input stream is delegated to the decoder implementation.
```
[the remark this answers is not in the document]
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/QueryDataDecoder.java:35]: No, decoder should close

**Comment by wendigo** [comment]
Benchmarks: fetching and decoding 4M rows over spooled protocol locally.

Before:
```
Average time for 20 runs: 12648ms
```

After:
```
Average time for 20 runs: 9361ms
```

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    @Override
+    public Iterable<List<Object>> toIterable()
+            throws IOException
+    {
+        return new RowWiseIterator(stream, decoders);
+    }
+
+    private static class RowWiseIterator
+            implements Iterable<List<Object>>, Iterator<List<Object>>
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:63]: This is not a valid implementation of Iterable. You cannot call `iterator()` method more than once and get what you would expect.

Comment (losipiuk) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:63]: If what you really need is an `Iterator`, not an `Iterable` change the interface of `QueryDataAccess`

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.currentToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_NUMBER_INT:
+                    return parser.getLongValue();
+                case VALUE_STRING:
+                    return Long.parseLong(parser.getValueAsString());
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:142]: was it like that before? Why do we have two on-wire representations?

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        public RowWiseIterator(InputStream stream, TypeDecoder[] decoders)
+                throws IOException
+        {
+            this.parser = JSON_FACTORY.createParser(stream);
+            this.decoders = requireNonNull(decoders, "decoders is null");
+            closer.register(stream);
+            closer.register(parser);
+            try {
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:79]: duplicate line

Comment (losipiuk) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:79]: Oh - that is intntiuonal :)
Please add some comment and maybe make error messages distinguishable

File: core/trino-main/src/test/java/io/trino/server/protocol/spooling/TestJsonQueryDataEncoding.java [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestJsonQueryDataEncoding
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/server/protocol/spooling/TestJsonQueryDataEncoding.java:72]: add some tests with multiple columns and some negative test, to verify exceptions

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 15]
```diff
@@ -103,6 +103,7 @@ protected QueryRunner createQueryRunner()
         postgreSqlServer = closeAfterClass(new TestingPostgreSqlServer());
         return PostgreSqlQueryRunner.builder(postgreSqlServer)
                 .setInitialTables(REQUIRED_TPCH_TABLES)
+                .withProtocolSpooling("json")
```
Comment (losipiuk) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:106]: would that make sense to run both versions?

Comment (wendigo) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:106]: Yeah, maybe so

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java (hunk 13 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:79]: Actually now, we want to test that JSON starts with `[[`

**Review by losipiuk — COMMENTED** [review]
LGTM % comment about iterable

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java (hunk 11 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:63]: I'll make a follow up PR - this is a bigger change

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java (hunk 12 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java:142]: Yes, you can check `FixJsonUtils`:

```
case BIGINT:
                return (value) -> {
                    if (value instanceof String) {
                        return Long.parseLong((String) value);
                    }
                    return ((Number) value).longValue();
                };
```

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java (hunk 11 above)
Comment (losipiuk) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java:63]: makes sense 👍

**Review by losipiuk — APPROVED** [review]

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java (hunk 2 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java]: Applied :)
